src/swaps/liquid_swapper.rs

```diff
@@ -0,0 +1,992 @@
+
+        let taproot_spend_info = taproot_builder.finalize(&secp, internal_key).unwrap(); // Taproot finalizing in unfailable
+
+        // Verify taproot construction
+        let output_key = taproot_spend_info.output_key();
+
+        let lockup_spk = Address::from_str(&create_swap_response.address)?.script_pubkey();
+
+        let pubkey_instruction = lockup_spk
+            .instructions()
+            .last()
+            .expect("should contain value")
+            .expect("should not fail");
+
+        let lockup_xonly_pubkey_bytes = pubkey_instruction
+            .push_bytes()
+            .expect("pubkey bytes expected");
+
+        let lockup_xonly_pubkey = XOnlyPublicKey::from_slice(lockup_xonly_pubkey_bytes)?;
+
+        debug_assert!(lockup_xonly_pubkey == output_key.into_inner());
+
+        log::info!("Taproot creation and verification success!");
+
+        // Step 3: Tweak the Key Cache with Taproot tweak
+        let tweak = taproot_spend_info.tap_tweak();
+
+        let tweaked_pubkey = key_agg_cache
+            .pubkey_xonly_tweak_add(&secp, SecretKey::from_slice(&tweak.to_byte_array())?)?;
+
+        debug_assert!(output_key.into_inner() == tweaked_pubkey.x_only_public_key().0);
+
+        log::info!("Musig2 tweaking success!");
+
+        let session_id = MusigSessionId::new(&mut thread_rng());
+
+        let msg = Message::from_digest_slice(
+            &Vec::from_hex(&claim_tx_response.transaction_hash).unwrap(),
+        )?;
+
+        // Step 4: Start the Musig2 Signing session
+        let mut extra_rand = [0u8; 32];
+        OsRng.fill_bytes(&mut extra_rand);
+
+        let (sec_nonce, pub_nonce) = key_agg_cache.nonce_gen(
+            &secp,
+            session_id,
+            self.keypair.public_key(),
+            msg,
+            Some(extra_rand),
+        )?;
+
+        let boltz_nonce = MusigPubNonce::from_slice(&Vec::from_hex(&claim_tx_response.pub_nonce)?)?;
+
+        let agg_nonce = MusigAggNonce::new(&secp, &[boltz_nonce, pub_nonce]);
+
+        let musig_session = MusigSession::new(&secp, &key_agg_cache, agg_nonce, msg);
+
+        let partial_sig =
+            musig_session.partial_sign(&secp, sec_nonce, &self.keypair, &key_agg_cache)?;
+
+        let is_partial_sig_valid = musig_session.partial_verify(
+            &secp,
+            &key_agg_cache,
+            partial_sig,
+            pub_nonce,
+            self.keypair.public_key(),
+        );
+
+        debug_assert!(is_partial_sig_valid == true);
+
+        log::info!("Partial Signature creation and verification success.");
+
+        Ok((partial_sig, pub_nonce))
+    }
+
+    /// Creates a refund timelocked transaction for subamrine swap.
+    pub fn submarine_refund(
+        &self,
+        create_swap_response: &CreateSwapResponse,
+        destination: Address,
+        fee: u64,
+    ) -> Result<Transaction, Error> {
+        // Step 1: Start with a Musig KeyAgg Cache
```

Comment:
Would be nice to put the creation and verification of the tree in a function to avoid code duplication